assets/src/stories-editor/components/right-click-menu/index.js

```diff
@@ -0,0 +1,271 @@
+/**
+ * External dependencies
+ */
+import { castArray } from 'lodash';
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { cloneBlock, pasteHandler, serialize } from '@wordpress/blocks';
+import { useEffect, useState } from '@wordpress/element';
+import {
+	MenuGroup,
+	MenuItem,
+	NavigableMenu,
+	Popover,
+} from '@wordpress/components';
+import { withDispatch, withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import './edit.css';
+import {
+	copyTextToClipBoard,
+	ensureAllowedBlocksOnPaste,
+} from '../../helpers';
+
+const POPOVER_PROPS = {
+	className: 'amp-story-right-click-menu__popover block-editor-block-settings-menu__popover editor-block-settings-menu__popover',
+	position: 'bottom left',
+};
+
+const RightClickMenu = ( props ) => {
+	const {
+		clientIds,
+		clientX,
+		clientY,
+		copyBlock,
+		cutBlock,
+		getBlock,
+		removeBlock,
+		duplicateBlock,
+		pasteBlock,
+	} = props;
+	const [ isOpen, setIsOpen ] = useState( true );
+	const [ copiedBlock, setCopiedBlock ] = useState( '' );
```

Comment:
As @westonruter brought up over the last grooming call then it's not possible to just read the clipboard data without and actual paste event.

WDYT of this workaround? Overview:
- When a block is either copied or cut, the block's markup is saved to the state as well.
- When something is saved to the state, and the right-click menu is opened, the Paste button will show up as well.
- When state is empty, the button is not displayed.

Shortcomings:
- When something else (e.g. a text or something from a different browser tab etc.) is copied meanwhile with the default browser (or keyboard) Copy then the state will still display the previously copied block. This could have a workaround of saving the copied block to a store instead of state, and we catch all the Copy events and empty the store where relevant.
- After refresh, the state is empty -- this means that pasting is not possible anymore with the button.
- Pasting to other Pages is not possible.

Note: Using <kbd><kbd>Ctrl</kbd>+<kbd>V</kbd></kbd> would still work as normally.

RightClick menu for a Block with sth saved to state:
<img width="357" alt="Screenshot 2019-08-29 at 14 05 31" src="https://user-images.githubusercontent.com/3294597/63938871-26ec3500-ca66-11e9-9b9a-70333a929417.png">


RightClick menu for Page:
<img width="360" alt="Screenshot 2019-08-29 at 14 05 23" src="https://user-images.githubusercontent.com/3294597/63938855-1b990980-ca66-11e9-805c-249fee68ea04.png">

Thoughts?

cc @swissspidy

Reply:
Nice idea! As you say, there are some shortcomings, but I think it's a good compromise and better than not having that option at all.

I'd probably use a store rather than state though for this. Otherwise 👍

Reply:
> it's not possible to just read the clipboard data without an actual paste event.

Looks like one just needs to use the Permissions API with the `clipboard-read` permission (Chrome only unfortunately)

https://developers.google.com/web/updates/2018/03/clipboardapi

The first time the user clicks on the paste button, we could ask for permission to directly use the clipboard:

https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read#Example

If permission is denied, we fall back to the data in the store/state.

Reply:
Yep, saw that it's possible in Chrome only.

Thoughts that this would cause some inconsistency in behavior though. E.g. in Chrome being able to copy from one tab and paste to the other and in other browsers not.

If that's acceptable then let's go for it. Thoughts?

Reply:
To me that sounds like progressive enhancement, so we should do it :-)